Bridge raising access denied exception when missing `content-type` in the request

When an operation against the bridge is executed without specifying the `content-type` in the header, the following error is logged in the pod and an "Internal Server Error" string is returned to the HTTP client.

```shell
2022-02-16 13:38:54 ERROR RoutingContext: - Unhandled exception in router
io.vertx.core.file.FileSystemException: Unable to create folder at path 'file-uploads'
	at io.vertx.core.file.impl.FileSystemImpl$12.perform(FileSystemImpl.java:950) ~[vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.file.impl.FileSystemImpl$12.perform(FileSystemImpl.java:932) ~[vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.file.impl.FileSystemImpl.mkdirsBlocking(FileSystemImpl.java:345) ~[vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.makeUploadDir(BodyHandlerImpl.java:248) ~[vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.<init>(BodyHandlerImpl.java:186) ~[vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl.handle(BodyHandlerImpl.java:84) ~[vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.handler.impl.BodyHandlerImpl.handle(BodyHandlerImpl.java:44) ~[vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212) ~[vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163) [vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141) [vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:67) [vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:37) [vertx-web-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:67) [vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xServerRequestHandler.handle(Http1xServerRequestHandler.java:30) [vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50) [vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.impl.DuplicatedContext.emit(DuplicatedContext.java:168) [vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:145) [vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:156) [vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153) [vertx-core-4.2.4.jar:4.2.4]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99) [netty-codec-http-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:116) [vertx-core-4.2.4.jar:4.2.4]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327) [netty-codec-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:314) [netty-codec-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:435) [netty-codec-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:279) [netty-codec-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61) [vertx-core-4.2.4.jar:4.2.4]
	at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38) [vertx-core-4.2.4.jar:4.2.4]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496) [netty-transport-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.73.Final.jar:4.1.73.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.73.Final.jar:4.1.73.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.nio.file.AccessDeniedException: /opt/strimzi/file-uploads
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:90) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111) ~[?:?]
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116) ~[?:?]
	at sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:389) ~[?:?]
	at java.nio.file.Files.createDirectory(Files.java:690) ~[?:?]
	at java.nio.file.Files.createAndCheckIsDirectory(Files.java:797) ~[?:?]
	at java.nio.file.Files.createDirectories(Files.java:783) ~[?:?]
	at io.vertx.core.file.impl.FileSystemImpl$12.perform(FileSystemImpl.java:940) ~[vertx-core-4.2.4.jar:4.2.4]
	... 55 more
```

The above error was raised by issuing the following example request via `cURL` where the `content-type` is missing.

```shell
curl -X POST \
  http://localhost:8080/topics/my-topic \
  -d '{
    "records": [
        {
            "key": "key-1",
            "value": "value-1"
        },
        {
            "key": "key-2",
            "value": "value-2"
        }
    ]
}'
```

By adding the line `-H 'content-type: application/vnd.kafka.json.v2+json'` the operation works fine.

My feeling is that the Vert.x router is defaulting to some "built-in" route which tries to create a `file-uploads` folder in a read-only `/opt/strimzi/` path on the pod, when it's not able to route the request properly due to the missing `content-type` because how it's defined in the OpenAPI specification.

--- a/src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java
+++ b/src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java
@@ -33,6 +33,7 @@ import io.vertx.ext.web.RoutingContext;
 
 import io.vertx.ext.web.handler.CorsHandler;
 import io.vertx.ext.web.openapi.RouterBuilder;
+import io.vertx.ext.web.validation.BodyProcessorException;
 import io.vertx.ext.web.validation.ParameterProcessorException;
 import io.vertx.json.schema.ValidationException;
 import org.apache.kafka.clients.CommonClientConfigs;
@@ -574,13 +575,20 @@ public class HttpBridge extends AbstractVerticle {
             }
 
             if (routingContext.failure() != null && routingContext.failure() instanceof ParameterProcessorException) {
-                ParameterProcessorException validationException = (ParameterProcessorException) routingContext.failure();
+                ParameterProcessorException parameterException = (ParameterProcessorException) routingContext.failure();
                 StringBuilder sb = new StringBuilder();
-                if (validationException.getParameterName() != null) {
-                    sb.append("Validation error on: ").append(validationException.getParameterName()).append(" - ");
+                if (parameterException.getParameterName() != null) {
+                    sb.append("Parameter error on: ").append(parameterException.getParameterName()).append(" - ");
                 }
-                sb.append(validationException.getMessage());
+                sb.append(parameterException.getMessage());
                 message = sb.toString();
+            }
+
+            if (routingContext.failure() != null && routingContext.failure() instanceof BodyProcessorException) {
+                BodyProcessorException bodyProcessorException = (BodyProcessorException) routingContext.failure();
+                if (bodyProcessorException.getMessage() != null) {
+                    message = bodyProcessorException.getMessage();
+                }
             }
         } else if (routingContext.statusCode() == HttpResponseStatus.NOT_FOUND.code()) {
             message = HttpResponseStatus.NOT_FOUND.reasonPhrase();

--- a/src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java
+++ b/src/main/java/io/strimzi/kafka/bridge/http/HttpBridge.java
@@ -564,31 +564,25 @@ public class HttpBridge extends AbstractVerticle {
         if (routingContext.statusCode() == HttpResponseStatus.BAD_REQUEST.code()) {
             message = HttpResponseStatus.BAD_REQUEST.reasonPhrase();
             // in case of validation exception, building a meaningful error message
-            if (routingContext.failure() != null && routingContext.failure().getCause() instanceof ValidationException) {
-                ValidationException validationException = (ValidationException) routingContext.failure().getCause();
+            if (routingContext.failure() != null) {
                 StringBuilder sb = new StringBuilder();
-                if (validationException.inputScope() != null) {
-                    sb.append("Validation error on: ").append(validationException.inputScope()).append(" - ");
+                if (routingContext.failure().getCause() instanceof ValidationException) {
+                    ValidationException validationException = (ValidationException) routingContext.failure().getCause();
+                    if (validationException.inputScope() != null) {
+                        sb.append("Validation error on: ").append(validationException.inputScope()).append(" - ");
+                    }
+                    sb.append(validationException.getMessage());
+                } else if (routingContext.failure() instanceof ParameterProcessorException) {
+                    ParameterProcessorException parameterException = (ParameterProcessorException) routingContext.failure();
+                    if (parameterException.getParameterName() != null) {
+                        sb.append("Parameter error on: ").append(parameterException.getParameterName()).append(" - ");
+                    }
+                    sb.append(parameterException.getMessage());
+                } else if (routingContext.failure() instanceof BodyProcessorException) {
+                    BodyProcessorException bodyProcessorException = (BodyProcessorException) routingContext.failure();
+                    sb.append(bodyProcessorException.getMessage());
                 }
-                sb.append(validationException.getMessage());
                 message = sb.toString();
-            }
-
-            if (routingContext.failure() != null && routingContext.failure() instanceof ParameterProcessorException) {
-                ParameterProcessorException parameterException = (ParameterProcessorException) routingContext.failure();
-                StringBuilder sb = new StringBuilder();
-                if (parameterException.getParameterName() != null) {
-                    sb.append("Parameter error on: ").append(parameterException.getParameterName()).append(" - ");
-                }
-                sb.append(parameterException.getMessage());
-                message = sb.toString();
-            }
-
-            if (routingContext.failure() != null && routingContext.failure() instanceof BodyProcessorException) {
-                BodyProcessorException bodyProcessorException = (BodyProcessorException) routingContext.failure();
-                if (bodyProcessorException.getMessage() != null) {
-                    message = bodyProcessorException.getMessage();
-                }
             }
         } else if (routingContext.statusCode() == HttpResponseStatus.NOT_FOUND.code()) {
             message = HttpResponseStatus.NOT_FOUND.reasonPhrase();

--- a/src/test/java/io/strimzi/kafka/bridge/http/ProducerIT.java
+++ b/src/test/java/io/strimzi/kafka/bridge/http/ProducerIT.java
@@ -939,4 +939,37 @@ public class ProducerIT extends HttpBridgeITAbstract {
 
         assertThat(context.awaitCompletion(TEST_TIMEOUT, TimeUnit.SECONDS), is(true));
     }
+
+    @Test
+    void sendSimpleMessageWithWrongContentType(VertxTestContext context) throws InterruptedException, ExecutionException {
+        KafkaFuture<Void> future = adminClientFacade.createTopic(topic);
+
+        String value = "message-value";
+
+        JsonArray records = new JsonArray();
+        JsonObject json = new JsonObject();
+        json.put("value", value);
+        records.add(json);
+
+        JsonObject root = new JsonObject();
+        root.put("records", records);
+
+        future.get();
+
+        producerService()
+                .sendRecordsRequest(topic, root, "bad-content-type")
+                .sendJsonObject(root, ar -> {
+                    context.verify(() -> {
+                        assertThat(ar.succeeded(), is(true));
+                        HttpResponse<JsonObject> response = ar.result();
+                        HttpBridgeError error = HttpBridgeError.fromJson(response.body());
+                        assertThat(response.statusCode(), is(HttpResponseStatus.BAD_REQUEST.code()));
+                        assertThat(error.getCode(), is(HttpResponseStatus.BAD_REQUEST.code()));
+                        assertThat(error.getMessage(), containsString("Cannot find body processor for content type"));
+                    });
+                    context.completeNow();
+                });
+
+        assertThat(context.awaitCompletion(TEST_TIMEOUT, TimeUnit.SECONDS), is(true));
+    }
 }

--- a/src/test/java/io/strimzi/kafka/bridge/http/ProducerIT.java
+++ b/src/test/java/io/strimzi/kafka/bridge/http/ProducerIT.java
@@ -647,7 +647,7 @@ public class ProducerIT extends HttpBridgeITAbstract {
 
         producerService()
             .sendRecordsToPartitionRequest(topic, partition, root, BridgeContentType.KAFKA_JSON_JSON)
-            .sendJsonObject(root, verifyBadRequest(context, "Validation error on: partitionid - [Bad Request] Parsing error for parameter partitionid in location PATH: java.lang.NumberFormatException: For input string: \"" + partition + "\""));
+            .sendJsonObject(root, verifyBadRequest(context, "Parameter error on: partitionid - [Bad Request] Parsing error for parameter partitionid in location PATH: java.lang.NumberFormatException: For input string: \"" + partition + "\""));
     }
 
     @Test

